pkg/slayers/path/scion/raw_spec.gobra

```diff
@@ -850,38 +844,52 @@ func XoverRightSeg(raw []byte, currInfIdx int, currHfIdx int, segs io.SegLens) {
 }
 
 ghost
-opaque
+requires  R55 < p
 requires  0 <= offset
-requires  0 <= currHFIdx && currHFIdx <= segLen
-requires  len(hops) == segLen - currHFIdx
+requires  0 <= currHfIdx && currHfIdx <= end
+requires  end <= segLen
 requires  offset + path.HopLen * segLen <= len(raw)
-requires  acc(sl.Bytes(raw, 0, len(raw)), R56)
-decreases
-pure func hopFieldsBytePositions(raw []byte, offset int, currHFIdx int, segLen int, hops seq[io.IO_HF]) bool {
-	return forall i int :: { hops[i] } 0 <= i && i < len(hops) ==>
-		hops[i] == path.BytesToIO_HF(raw, 0, offset + path.HopLen * (currHFIdx + i), len(raw))
+preserves acc(sl.Bytes(raw, 0, len(raw)), p)
+ensures   hopFields(raw, offset, currHfIdx, segLen)[:end-currHfIdx] ==
+	hopFields(raw, offset, currHfIdx, end)
+decreases end - currHfIdx
+func hopFieldsSliceEnd(raw []byte, offset int, currHfIdx int, segLen int, end int, p perm) {
```

Comment:
I would prefer a different naming scheme for lemmas that describes the result that is established. Here, for example, I would have gone with `HopsFromPrefixOfRawMatchPrefixOfHops` or something like that. This applies to all lemmas introduced here. Also, I could see this being useful for clients, so I would make this public (I wonder if we already have some lemmas in other packages that could be useful here).

Reply:
We have similar functions, such as `widenHopFields()`, but they aren't applicable in this case. I attempted to consolidate these similar lemmas into a single, more general lemma. However, doing so proved to be quite tricky and made the resulting code difficult to read. Consequently, I decided to create multiple, more specific lemmas instead.